File: src/test/java/com/ruoyi/asset/domain/enums/AssetStatusEnumTransitionTest.java
package com.ruoyi.asset.domain.enums;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.junit.MockitoJUnitRunner;

import java.util.Arrays;
import java.util.List;

import static org.junit.Assert.*;

/**
 * AssetStatusEnum Transition Test
 *
 * Tests for AssetStatusEnum state transitions and validation
 *
 * @author ruoyi
 * @date 2025-12-15
 */
@RunWith(MockitoJUnitRunner.class)
public class AssetStatusEnumTransitionTest {

    @Test
    public void shouldDefineAllRequiredStatusValuesInAssetStatusEnum() {
        // Test that all required status values exist
        List<AssetStatusEnum> statusValues = Arrays.asList(
            AssetStatusEnum.UNDER_CONSTRUCTION,
            AssetStatusEnum.AVAILABLE_SELF,
            AssetStatusEnum.AVAILABLE_RENT,
            AssetStatusEnum.AVAILABLE_IDLE,
            AssetStatusEnum.TEMP_CLOSED,
            AssetStatusEnum.FAULT,
            AssetStatusEnum.MAINTAINING,
            AssetStatusEnum.TO_BE_SCRAPPED,
            AssetStatusEnum.SCRAPPED
        );

        assertNotNull("UNDER_CONSTRUCTION status should exist", AssetStatusEnum.UNDER_CONSTRUCTION);
        assertNotNull("AVAILABLE_SELF status should exist", AssetStatusEnum.AVAILABLE_SELF);
        assertNotNull("AVAILABLE_RENT status should exist", AssetStatusEnum.AVAILABLE_RENT);
        assertNotNull("AVAILABLE_IDLE status should exist", AssetStatusEnum.AVAILABLE_IDLE);
        assertNotNull("TEMP_CLOSED status should exist", AssetStatusEnum.TEMP_CLOSED);
        assertNotNull("FAULT status should exist", AssetStatusEnum.FAULT);
        assertNotNull("MAINTAINING status should exist", AssetStatusEnum.MAINTAINING);
        assertNotNull("TO_BE_SCRAPPED status should exist", AssetStatusEnum.TO_BE_SCRAPPED);
        assertNotNull("SCRAPPED status should exist", AssetStatusEnum.SCRAPPED);

        assertEquals("Should have 9 status values", 9, statusValues.size());
    }

    @Test
    public void shouldHaveTransitionValidationMapConfiguredForEachStatus() {
        // Test that each status has valid transitions configured
        for (AssetStatusEnum status : AssetStatusEnum.values()) {
            assertNotNull("Status " + status + " should have valid transitions",
                status.getValidTransitions());
        }

        // SCRAPPED should have no valid transitions (terminal state)
        assertNotNull("SCRAPPED should have valid transitions (empty list)",
            AssetStatusEnum.SCRAPPED.getValidTransitions());
        assertTrue("SCRAPPED should have empty valid transitions",
            AssetStatusEnum.SCRAPPED.getValidTransitions().isEmpty());
    }

    @Test
    public void shouldReturnTrueForValidStateTransitionsViaCanTransitionTo() {
        // Test valid transitions
        assertTrue("UNDER_CONSTRUCTION should transition to AVAILABLE_SELF",
            AssetStatusEnum.UNDER_CONSTRUCTION.canTransitionTo(AssetStatusEnum.AVAILABLE_SELF));
        assertTrue("AVAILABLE_SELF should transition to MAINTAINING",
            AssetStatusEnum.AVAILABLE_SELF.canTransitionTo(AssetStatusEnum.MAINTAINING));
        assertTrue("AVAILABLE_RENT should transition to FAULT",
            AssetStatusEnum.AVAILABLE_RENT.canTransitionTo(AssetStatusEnum.FAULT));
        assertTrue("AVAILABLE_IDLE should transition to TEMP_CLOSED",
            AssetStatusEnum.AVAILABLE_IDLE.canTransitionTo(AssetStatusEnum.TEMP_CLOSED));
        assertTrue("MAINTAINING should transition to AVAILABLE_SELF",
            AssetStatusEnum.MAINTAINING.canTransitionTo(AssetStatusEnum.AVAILABLE_SELF));
        assertTrue("TO_BE_SCRAPPED should transition to SCRAPPED",
            AssetStatusEnum.TO_BE_SCRAPPED.canTransitionTo(AssetStatusEnum.SCRAPPED));
    }

    @Test
    public void shouldReturnFalseForInvalidStateTransitionsViaCanTransitionTo() {
        // Test invalid transitions
        assertFalse("SCRAPPED should not transition to AVAILABLE_SELF",
            AssetStatusEnum.SCRAPPED.canTransitionTo(AssetStatusEnum.AVAILABLE_SELF));
        assertFalse("SCRAPPED should not transition to UNDER_CONSTRUCTION",
            AssetStatusEnum.SCRAPPED.canTransitionTo(AssetStatusEnum.UNDER_CONSTRUCTION));
        assertFalse("SCRAPPED should not transition to MAINTAINING",
            AssetStatusEnum.SCRAPPED.canTransitionTo(AssetStatusEnum.MAINTAINING));
        assertFalse("UNDER_CONSTRUCTION should not transition to SCRAPPED directly",
            AssetStatusEnum.UNDER_CONSTRUCTION.canTransitionTo(AssetStatusEnum.SCRAPPED));
    }

    @Test
    public void shouldReturnCorrectSetOfValidTransitionsViaGetValidTransitions() {
        // Test getValidTransitions method
        assertNotNull("AVAILABLE_SELF should have valid transitions",
            AssetStatusEnum.AVAILABLE_SELF.getValidTransitions());
        assertTrue("AVAILABLE_SELF should have valid transitions",
            AssetStatusEnum.AVAILABLE_SELF.getValidTransitions().size() > 0);

        assertNotNull("SCRAPPED should have valid transitions",
            AssetStatusEnum.SCRAPPED.getValidTransitions());
        assertEquals("SCRAPPED should have empty valid transitions",
            0, AssetStatusEnum.SCRAPPED.getValidTransitions().size());

        // Verify that valid transitions include expected statuses
        assertTrue("AVAILABLE_SELF transitions should include MAINTAINING",
            AssetStatusEnum.AVAILABLE_SELF.getValidTransitions().contains(AssetStatusEnum.MAINTAINING));
        assertTrue("MAINTAINING transitions should include AVAILABLE_SELF",
            AssetStatusEnum.MAINTAINING.getValidTransitions().contains(AssetStatusEnum.AVAILABLE_SELF));
    }

    @Test
    public void shouldHaveCorrectStatusCodeAndDescription() {
        // Test that each status has code and description
        for (AssetStatusEnum status : AssetStatusEnum.values()) {
            assertNotNull("Status " + status + " should have code",
                status.getCode());
            assertNotNull("Status " + status + " should have description",
                status.getDescription());
            assertFalse("Status " + status + " code should not be empty",
                status.getCode().isEmpty());
            assertFalse("Status " + status + " description should not be empty",
                status.getDescription().isEmpty());
        }
    }

    @Test
    public void shouldFindStatusByCode() {
        // Test fromCode method
        assertEquals("Should find UNDER_CONSTRUCTION by code",
            AssetStatusEnum.UNDER_CONSTRUCTION, AssetStatusEnum.fromCode("under_construction"));
        assertEquals("Should find AVAILABLE_SELF by code",
            AssetStatusEnum.AVAILABLE_SELF, AssetStatusEnum.fromCode("available_self"));
        assertEquals("Should find SCRAPPED by code",
            AssetStatusEnum.SCRAPPED, AssetStatusEnum.fromCode("scrapped"));
    }

    @Test
    public void shouldReturnNullForInvalidStatusCode() {
        // Test fromCode with invalid code
        assertNull("Should return null for invalid code",
            AssetStatusEnum.fromCode("INVALID_STATUS"));
        assertNull("Should return null for null code",
            AssetStatusEnum.fromCode(null));
        assertNull("Should return null for empty code",
            AssetStatusEnum.fromCode(""));
    }
}

File: tests/core/asset-status-enum-transition.test.ts
import { describe, it, expect } from 'vitest';

// Mock Java enum with transition logic
const createStatus = (code: string, description: string, transitions: string[]) => {
  const statuses: Record<string, any> = {
    UNDER_CONSTRUCTION: ['AVAILABLE_SELF', 'AVAILABLE_RENT', 'AVAILABLE_IDLE'],
    AVAILABLE_SELF: ['MAINTAINING', 'FAULT', 'TEMP_CLOSED', 'TO_BE_SCRAPPED'],
    AVAILABLE_RENT: ['MAINTAINING', 'FAULT', 'TEMP_CLOSED', 'TO_BE_SCRAPPED'],
    AVAILABLE_IDLE: ['MAINTAINING', 'FAULT', 'TEMP_CLOSED', 'TO_BE_SCRAPPED'],
    TEMP_CLOSED: ['AVAILABLE_SELF', 'AVAILABLE_RENT', 'AVAILABLE_IDLE'],
    FAULT: ['AVAILABLE_SELF', 'AVAILABLE_RENT', 'AVAILABLE_IDLE', 'TO_BE_SCRAPPED'],
    MAINTAINING: ['AVAILABLE_SELF', 'AVAILABLE_RENT', 'AVAILABLE_IDLE'],
    TO_BE_SCRAPPED: ['SCRAPPED', 'AVAILABLE_SELF', 'AVAILABLE_RENT', 'AVAILABLE_IDLE'],
    SCRAPPED: []
  };

  return {
    code,
    description,
    canTransitionTo(target: any) {
      const validTransitions = statuses[this.name];
      return validTransitions.includes(target.name);
    },
    getValidTransitions() {
      const validTransitions = statuses[this.name];
      return validTransitions.map((name: string) => AssetStatusEnum[name]);
    }
  };
};

const AssetStatusEnum: Record<string, any> = {
  UNDER_CONSTRUCTION: Object.assign(createStatus('under_construction', '在建/未投入'), { name: 'UNDER_CONSTRUCTION' }),
  AVAILABLE_SELF: Object.assign(createStatus('available_self', '可用-自用'), { name: 'AVAILABLE_SELF' }),
  AVAILABLE_RENT: Object.assign(createStatus('available_rent', '可用-对外出租'), { name: 'AVAILABLE_RENT' }),
  AVAILABLE_IDLE: Object.assign(createStatus('available_idle', '可用-空置'), { name: 'AVAILABLE_IDLE' }),
  TEMP_CLOSED: Object.assign(createStatus('temp_closed', '暂停使用'), { name: 'TEMP_CLOSED' }),
  FAULT: Object.assign(createStatus('fault', '故障/停用'), { name: 'FAULT' }),
  MAINTAINING: Object.assign(createStatus('maintaining', '维修中'), { name: 'MAINTAINING' }),
  TO_BE_SCRAPPED: Object.assign(createStatus('to_be_scrapped', '计划报废'), { name: 'TO_BE_SCRAPPED' }),
  SCRAPPED: Object.assign(createStatus('scrapped', '已报废'), { name: 'SCRAPPED' })
};

describe('AssetStatusEnum', () => {
  it('should have 9 status values', () => {
    const statusValues = Object.keys(AssetStatusEnum);
    expect(statusValues).toHaveLength(9);
  });

  it('should validate status transitions', () => {
    // Test valid transitions
    expect(AssetStatusEnum.AVAILABLE_IDLE.canTransitionTo(AssetStatusEnum.TEMP_CLOSED)).toBe(true);
    expect(AssetStatusEnum.TEMP_CLOSED.canTransitionTo(AssetStatusEnum.AVAILABLE_IDLE)).toBe(true);
    expect(AssetStatusEnum.AVAILABLE_IDLE.canTransitionTo(AssetStatusEnum.FAULT)).toBe(true);

    // Test invalid transitions (SCRAPPED is terminal)
    expect(AssetStatusEnum.SCRAPPED.canTransitionTo(AssetStatusEnum.AVAILABLE_IDLE)).toBe(false);
    expect(AssetStatusEnum.SCRAPPED.canTransitionTo(AssetStatusEnum.UNDER_CONSTRUCTION)).toBe(false);

    // Test UNDER_CONSTRUCTION can only go to AVAILABLE_* states
    expect(AssetStatusEnum.UNDER_CONSTRUCTION.canTransitionTo(AssetStatusEnum.AVAILABLE_SELF)).toBe(true);
    expect(AssetStatusEnum.UNDER_CONSTRUCTION.canTransitionTo(AssetStatusEnum.AVAILABLE_RENT)).toBe(true);
    expect(AssetStatusEnum.UNDER_CONSTRUCTION.canTransitionTo(AssetStatusEnum.AVAILABLE_IDLE)).toBe(true);
  });

  it('should return valid transitions', () => {
    const validTransitions = AssetStatusEnum.AVAILABLE_IDLE.getValidTransitions();
    expect(validTransitions).toContain(AssetStatusEnum.TEMP_CLOSED);
    expect(validTransitions).toContain(AssetStatusEnum.FAULT);
    expect(validTransitions).toContain(AssetStatusEnum.MAINTAINING);
  });

  it('should have terminal state with no valid transitions', () => {
    const validTransitions = AssetStatusEnum.SCRAPPED.getValidTransitions();
    expect(validTransitions).toHaveLength(0);
  });
});
